Add footer

GuillaumeGomez opened the pull request:
Fixes #75 

![Screenshot from 2021-04-21 14-55-57](https://user-images.githubusercontent.com/3050060/115557888-7114d080-a2b2-11eb-8079-3fd35970c028.png)
![Screenshot from 2021-04-21 15-28-09](https://user-images.githubusercontent.com/3050060/115561945-517fa700-a2b6-11eb-83bc-fc7f07a3246a.png)

As you can see, whatever the size of the page, it's always at the bottom and not in the middle of the screen but if the content if bigger that the height, you have to scroll to the bottom of the page to be able to see the footer.

Little sidenode for the rustdoc pages in mobile mode. In such case, the footer takes the full width and goes under the sidebar in case it's displayed:

![Screenshot from 2021-04-21 15-27-56](https://user-images.githubusercontent.com/3050060/115562124-7c69fb00-a2b6-11eb-919a-ec085efe4d18.png)
![Screenshot from 2021-04-21 15-28-00](https://user-images.githubusercontent.com/3050060/115562130-7d029180-a2b6-11eb-8ce9-6c0c26e96851.png)

jyn514 commented:
I wonder if it makes sense to move the 'About' and homepage links in the bottom. 'About' because it seems common to put that in the footer, and the homepage because to be frank there's not much on our homepage other than search, which we already show in the top-right. That would also free up some space in the header.

jyn514 commented on templates/footer.html:
```diff
@@ -0,0 +1,3 @@
+<div class="docs-rs-footer">
+    Who wants some content?
```
```suggestion
    <a href="https://foundation.rust-lang.org/policies/privacy-policy/#docs.rs">Privacy</a>
```

GuillaumeGomez replied:
It can't be the only thing we will put in there, right? Do we also want to add something like "docs.rs is owned by the Rust foundation" or something along the line?

jyn514 replied:
I don't think we need to? I would only put things there that are useful things people might want to click, I don't think we should add links just to take up space.

GuillaumeGomez replied:
Ok!

Nemo157 commented on templates/style/rustdoc.scss:
```diff
@@ -1,13 +1,29 @@
 // FIXME: Use modules
 @import "vars", "navbar", "themes", "fa";
 
+body.rustdoc-age {
```
```suggestion
body.rustdoc-page {
```

I assume by the next rule?

GuillaumeGomez replied:
Woups, indeed!

Nemo157 commented:
Is there a reason to not make the footer fullwidth on rustdoc pages?

The other page it'd be useful to verify it works on is rustdoc-src pages, since they ~~don't have the sidebar~~ have a collapsible sidebar.

GuillaumeGomez commented:
@Nemo157 You'll see in a bit. ;)

GuillaumeGomez commented:
Updated the images and the PR. With this, I think it's now ready for review.

Nemo157 commented:
Ah, so the footer scrolls into view on long rustdoc pages, so that's why it can't be across both the page and sidebar. Still seems a bit weird in that it implies that it's part of the rustdoc output; maybe having some kind of `DOCS.RS` wordmark in the footer would help avoid that.

On rustdoc-src pages it doesn't extend all the way left:

![image](https://user-images.githubusercontent.com/81079/115267899-ffb21200-a139-11eb-9529-06cbfe60e007.png)

Nemo157 commented on templates/footer.html:
```diff
@@ -0,0 +1,4 @@
+<div class="docs-rs-footer">
+    <a href="https://foundation.rust-lang.org/policies/privacy-policy/#docs.rs">Privacy</a>
```
I think _Privacy_ on its own looks a bit weird, _Privacy policy_ would probably be better (and is what crates.io uses).

GuillaumeGomez replied:
:+1:

GuillaumeGomez commented:
Fixed the position on the source code page and updated "Privacy" to "Privacy policy".

pietroalbini commented on templates/footer.html:
```diff
@@ -0,0 +1,4 @@
+<div class="docs-rs-footer">
+    <a href="https://foundation.rust-lang.org/policies/privacy-policy/#docs.rs">Privacy policy</a>
+    <a href="https://foundation.rust-lang.org/">Rust Foundation</a>
```
Hmm, is there a reason to link to the foundation's home page here?

GuillaumeGomez replied:
Just policy seemed a bit weird. So I just said to myself: "why not?".

jyn514 replied:
@GuillaumeGomez I would rather you moved the 'about' links to the footer instead: https://github.com/rust-lang/docs.rs/pull/1367#issuecomment-822051232

GuillaumeGomez replied:
Fine by me! Gonna need a little reword on the top navbar resize then.

GuillaumeGomez replied:
After taking a look, I don't think we should move "about" from the topbar: it provides information that are required when you want to customize your "docs.rs experience". It's not "extra info" like what we put in the footer.

GuillaumeGomez replied:
If the issue is the "Rust foundation" link, I can simply remove it. :)

pietroalbini replied:
Honestly I would remove both the "About" and "Releases" links from the navbar, and have the footer be:

```
About docs.rs | Privacy Policy | Build queue
```

I looked at the stats we have on Grafana for the past 7 days:

* The whole about section got 0.0045% of the visits
* The whole release section except for the build queue got 0.0023% of the visits
* The build queue page got 0.0008% of the visits

I don't think either of the links in the navbar are that used to warrant their positioning.

GuillaumeGomez replied:
Ok, let's go with that then.

pietroalbini commented:
Another thing I noticed is that the footer on rustdoc pages uses the rustdoc serif font, could it be switched to use docs.rs's main font?

GuillaumeGomez commented:
@pietroalbini Wow, you have good eye. Didn't spot this one at all!

GuillaumeGomez commented:
Updated the links, added icons and updated the screenshots.

jyn514 commented on templates/footer.html:
```diff
@@ -0,0 +1,14 @@
+        <span title="About">{{ "info-circle" | fas }}</span>
+        <span class="title">About docs.rs</span>
+    </a>
+    <a href="https://foundation.rust-lang.org/policies/privacy-policy/#docs.rs">
+        <span title="Privacy">{{ "shield-alt" | fas }}</span>
+        <span class="title">Privacy policy</span>
```
Why add a separate span instead of using `<a title="Privacy policy>`?
```suggestion
    <a title="Privacy policy" href="https://foundation.rust-lang.org/policies/privacy-policy/#docs.rs">
        {{ "shield-alt" | fas }} Privacy policy
```

GuillaumeGomez replied:
It's done like this in the navbar so I replicated the same behavior.

pietroalbini commented:
Hmm, looking at the screenshots the icons feels too prominent? I think I prefer the icon-less version.

GuillaumeGomez commented:
So before removing them again, what do you think @rust-lang/docs-rs ? :)

GuillaumeGomez commented:
Removed the icons and updated the screenshots.

pietroalbini commented:
The font in rustdoc pages still seems to be serif. Otherwise that looks great!

GuillaumeGomez commented:
@pietroalbini You're absolutely right! I updated the font so it's now sans-serif. :)

Nemo157 commented:
Found one fun bug, the footer renders at the bottom of the initial page load and scrolls with the content on non-rustdoc pages.

![image](https://user-images.githubusercontent.com/81079/115572930-a0323e80-a2c0-11eb-9025-09579ea7da4a.png)

GuillaumeGomez commented:
Nice!

GuillaumeGomez commented:
So there were actually two issues:
1. CSS `position` is never `relative` by default (uuuurg) so I had to set it as such on `<body>`.
2. Some docs.rs pages were using `margin-bottom` or didn't have bottom margin at all, so I used `padding-bottom` in both cases, so on pages that didn't use bottom margin, the end of the page won't be right under the text. It's actually a nice improvement, I never paid attention to that before. :)

Nemo157 commented:
Looks like the about pages could do with a little extra `padding-bottom` too:

![image](https://user-images.githubusercontent.com/81079/116212421-19101b00-a745-11eb-94f8-e7e4a04ee5f6.png)

GuillaumeGomez commented:
I don't have the issue on the `/builds` page:

![Screenshot from 2021-04-27 10-50-53](https://user-images.githubusercontent.com/3050060/116214032-9ab47880-a746-11eb-95a0-c64c0532a8d5.png)

Nemo157 commented:
`/<crate>/builds` is different from `/about/builds`

GuillaumeGomez commented:
Indeed, also I confirmed that this issue only appeared on mobile.

GuillaumeGomez commented:
Fixed! So I applied @Nemo157's suggestion and added bottom padding on the about pages (which forced me to update the DOM and create a new CSS class to correctly handle it).

Nemo157 reviewed (approved):
Styling all looks good to me now, I don't have much of an opinion on what links are in it.

syphar reviewed (approved):
I did a round of tests, in Firefox/Chrome and in mobile sizes. 

All seems fine to me
